### PR TITLE
feat(providers): add JazzHR public career-page provider

### PR DESCRIPTION
**Pull request by ilyakanevskiy**
## Summary

Adds a zero-auth `jazzhr` provider for public `*.applytojob.com` career pages (JazzHR/ApplyToJob), auto-detected from `careers_url`/`api` on either the bare tenant host or `/apply`.

## Provenance

The original implementation of this provider was written by Schlaflied, in PR #3745. That GitHub account has since been deleted, and the PR no longer resolves (both it and its linked issue return 404 — deleted upstream, not just closed). I had reviewed that PR before it disappeared and had a cached copy of its final, twice-approved diff, so this is a reconstruction from that cache rather than a from-scratch reimplementation — same parsing approach, same guard shapes, full credit to Schlaflied for the original design. It is not a straight port: the card-recognition fallback below closes a real gap a reviewer flagged on the original PR and that never got fixed before it was deleted.

## What's included

- Host allowlist (anchored `*.applytojob.com` regex, HTTPS-only, `redirect: 'error'` on every request) before any network call. Only the hostname is trusted; the board URL is always rebuilt as `https://${hostname}/apply`, never carried through from whatever `careers_url`/`api` happened to contain — same shape `providers/bamboohr.mjs` / `providers/breezy.mjs` use for a subdomain-per-tenant ATS (hardcoding the scheme this way, rather than reusing `URL.origin`, is what makes the port irrelevant too — there's nothing left of the input for a port to survive in). A bare host, `/apply`, a trailing slash, a stray posting permalink, or a non-default port all resolve to the same board root; the path forms are genuinely in use in the wild (confirmed live — some companies' own careers pages link the bare host, others link `/apply` directly).
- Primary parse reads the `list-group-item` card wrapper (title + location together); a fallback anchors directly on the `/apply/` permalink when that wrapper is absent or renamed, so a markup redesign yields real postings instead of reading as a healthy empty board. The fallback runs unconditionally (not just when the primary pass finds nothing), so a partial redesign — some cards still wrapped, some not — doesn't silently drop the unwrapped ones.
- A board that carries `/apply/` links but parses to zero postings throws (markup broke); a board with none of that shape returns `[]` (genuinely empty) — the two are never conflated.
- Optional `jazzhr: { fetchDetails: true, detailLimit: N }` per-entry config fetches each posting's own page for its JSON-LD description and posted date, capped at `detailLimit` (default 25, max 100), paced at 200ms between requests, retried, and skipped entirely under a `ctx.maxPages` liveness probe.

## Docs / examples

- `docs/SUPPORTED_JOB_BOARDS.md` — added the JazzHR row in alphabetical position.
- `templates/portals.example.yml` — auto-detection pattern line, a commented `Example JazzHR Co` stanza with both config keys documented inline, and one live entry (Ratehub, Canada/Vancouver section).

## Test plan

- `node test-all.mjs --only providers/jazzhr` — 31/31 pass: origin-only board resolution (any path shape or port, incl. a posting permalink or `:8443`, lands on the board root), host/protocol allowlist incl. suffix-host and userinfo SSRF bypass attempts, list parsing with HTML entities, JSON-LD detail enrichment and its list-vs-detail location precedence, the wrapper-less fallback (incl. a partial-redesign mix of wrapped and unwrapped postings on the same page), empty-vs-broken board handling (incl. a query string or a bare `/apply//` collapsing onto the board root rather than being read as a posting), detail-loop pacing, and probe cooperation.
- `node test-all.mjs` — full suite green.
- Live-verified against 5 real tenants (read-only, no writes): onefootball.applytojob.com (5 postings), westernachersolutions.applytojob.com (17), holafly.applytojob.com (19), fairtiq.applytojob.com (6), ratehubinc.applytojob.com (15).

## Summary

Adds zero-auth JazzHR support for public HTTPS `*.applytojob.com` career pages.

Users can:
: Auto-detect boards from `api` or `careers_url` and normalize them to `/apply` (`providers/jazzhr.mjs:52-63`).
: Parse listing cards and fallback posting links (`providers/jazzhr.mjs:123-171`).
: Enable optional JSON-LD enrichment with `fetchDetails` and `detailLimit` (`providers/jazzhr.mjs:198-229`).
: Reject non-HTTPS hosts, unsafe hosts, and redirects (`providers/jazzhr.mjs:70-76`, `215-226`).

Invalid or unsupported JazzHR URLs do not produce a provider match. Detail-fetch failures do not remove valid list results.

Touched files:
: `providers/jazzhr.mjs`
: `tests/providers/jazzhr.test.mjs`
: `docs/SUPPORTED_JOB_BOARDS.md`
: `templates/portals.example.yml`

The change touches the `providers/` system area. It does not touch `AGENTS.md`, `modes/`, `update-system.mjs`, `DATA_CONTRACT.md`, or `.github/`.

The reported tests pass: 31 provider-specific tests and the full suite. Live verification covered five tenants.

One documentation issue remains: the provider header now correctly states that permalink fallback runs unconditionally (`providers/jazzhr.mjs:148-157`), but the review noted that the corresponding documentation still describes the old conditional behavior. The branch also needs a refresh against `main` and a new CI run.

### CODE REVIEW

**Comment by ilyakanevskiy** [comment]
@coderabbitai review

File: docs/SUPPORTED_JOB_BOARDS.md [hunk 1]
```diff
@@ -47,6 +47,7 @@ are shared helpers and are not loaded as providers.
...
+| JazzHR | Parser | Auto-detects public `*.applytojob.com` career pages at either the bare tenant host or `/apply`, commonly branded with `Powered by JazzHR`. Reads server-rendered listing cards and optional JSON-LD detail pages (`fetchDetails`). |
```
[the remark this answers is not in the document]
Comment (ilyakanevskiy) [reply docs/SUPPORTED_JOB_BOARDS.md:50]: Agreed this isn't a merge requirement, per your own note. Separately: tracking `career-ops-docs` staying in sync isn't an established convention in this repo yet — I don't see any existing PR or CI check here that gates on it, and `docs/SUPPORTED_JOB_BOARDS.md` itself is the local source of truth this checklist item asks for. Not doing anything here beyond what's already in this PR; happy to see that convention added as its own follow-up if a maintainer wants it.

File: providers/jazzhr.mjs [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+  }
+
+  // Fallback: `list-group-item` is a markup detail, not a JazzHR guarantee —
+  // a redesign that renames or drops that wrapper while keeping the /apply/
+  // permalink shape would otherwise read as a healthy empty board. Anchor
+  // directly on the posting link instead. Location isn't recoverable
+  // outside the card wrapper, so this pass only ever adds a title/url the
+  // primary pass missed — it never downgrades a result the primary pass
+  // already found.
+  if (jobs.length === 0) {
```
[the remark this answers is not in the document]
Comment (ilyakanevskiy) [reply providers/jazzhr.mjs]: Confirmed and fixed — a board with some cards still in the list-group-item wrapper and others already migrated off it was silently dropping the unwrapped ones with no signal, since the fallback only ran when the primary pass found nothing at all. Made it run unconditionally, as suggested; the shared `seen` set already made that safe (no duplicates). Added a regression test for the mixed-markup case and re-verified against all 5 live tenants this PR was tested against — identical job counts, nothing spurious added.

File: providers/jazzhr.mjs [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+      const url = resolvePostingUrl(match[1], base);
+      if (!url) continue;
+      addJob(url, clean(match[2]), '');
+      if (jobs.length >= MAX_JOBS) break;
+    }
+  }
+
+  if (jobs.length === 0 && /<a\b[^>]*href=["'][^"']*\/apply\//i.test(html)) {
+    throw new Error('jazzhr: found ApplyToJob links but could not parse any posting cards');
+  }
```
[the remark this answers is not in the document]
Comment (ilyakanevskiy) [reply providers/jazzhr.mjs:167]: Real concern, not implementing it right now. I don't have a verified sample of JazzHR's actual "documented empty template" markup to build a positive check against — your own research confirms there's no special empty-board marker, just the normal page shell with no listing elements, and I haven't fetched a genuinely-empty live tenant to see what that shell reliably contains across different customizations (branded/white-labeled boards in particular). A guessed signature risks throwing on a real healthy empty board, which is a worse failure mode for a scanning tool than the current gap — that gap has not manifested on any of this PR's 5 verified live tenants. This is the same tradeoff the original (now-deleted) PR shipped with; a reviewer flagged the identical concern there and it stayed open, non-blocking, through two approvals. Leaving it as a known limitation rather than shipping an unverified heuristic; open to a follow-up once there's a real empty-board sample to work from.

[the remark this answers is not in the document]
Comment (ilyakanevskiy) [reply providers/jazzhr.mjs:167]: No thanks, not worth a tracked issue for this — it's a narrow, non-blocking limitation on a single-company provider, not something with an active workload behind it. The comment in the code and this thread are enough of a record if anyone picks it up later.

File: providers/jazzhr.mjs [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+  if (typeof raw !== 'string' || !raw.trim()) return null;
+  try {
+    const url = new URL(raw.trim());
+    if (url.protocol !== 'https:' || !HOST_RE.test(url.hostname)) return null;
+    // Real JazzHR boards serve the same listing at the bare host and at
+    // /apply — every natural route in (a company's own careers button, an
+    // aggregator link, a search hit) hands over the bare host, never the
+    // constructed /apply form, so both must resolve.
+    if (url.pathname === '/' || url.pathname === '') url.pathname = '/apply';
+    else if (!/^\/apply(?:\/.*)?$/i.test(url.pathname)) return null;
```
[the remark this answers is not in the document]
Comment (ilyakanevskiy) [reply providers/jazzhr.mjs]: Confirmed and fixed — `parseBoardUrl`'s `/^\/apply(?:\/.*)?$/i` accepted any depth under `/apply/`, so a `careers_url` pointing at one specific posting instead of the board root would get fetched and parsed as the listing page. Split it: `parseBoardUrl` now only accepts the board root (`/apply` or `/apply/`, nothing after); `assertJazzHRUrl` got its own implementation that validates the deep `/apply/<id>/...` form instead, since that's what it's actually checking (a posting permalink before a detail fetch), not a board URL. Added a test asserting a posting-permalink `careers_url` is rejected, and re-verified `fetchDetails` still works end-to-end against a live tenant (assertJazzHRUrl still accepts the real posting URLs it's fed during the detail loop).

File: providers/jazzhr.mjs [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+ * path parseBoardUrl rejects, since this validates a specific job's URL,
+ * not the board root. */
+function assertJazzHRUrl(raw) {
+  let parsed;
+  try {
+    parsed = new URL(String(raw).trim());
+  } catch {
+    parsed = null;
+  }
+  if (!parsed || parsed.protocol !== 'https:' || !HOST_RE.test(parsed.hostname) || !/^\/apply\//i.test(parsed.pathname)) {
```
[the remark this answers is not in the document]
Comment (ilyakanevskiy) [reply providers/jazzhr.mjs]: Confirmed and fixed — both `assertJazzHRUrl` and `resolvePostingUrl` used `/^\/apply\//i`, which matches pathname `/apply/` itself (the board root) since nothing was required after the trailing slash. Concretely reachable via an href like `/apply/?ref=1`: the raw-string capture in the parsing regexes only requires *some* non-quote character after `/apply/`, which a query string satisfies, but `new URL()` splits that off into `.search`, leaving pathname exactly `/apply/`. Tightened both to `/^\/apply\/.+$/i`. Added a regression test for that exact href shape and re-verified against all 5 live tenants — same job counts as before.

File: providers/jazzhr.mjs [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+ * path parseBoardUrl rejects, since this validates a specific job's URL,
+ * not the board root. */
+function assertJazzHRUrl(raw) {
+  let parsed;
+  try {
+    parsed = new URL(String(raw).trim());
+  } catch {
+    parsed = null;
+  }
+  if (!parsed || parsed.protocol !== 'https:' || !HOST_RE.test(parsed.hostname) || !/^\/apply\/.+$/i.test(parsed.pathname)) {
```
[the remark this answers is not in the document]
Comment (ilyakanevskiy) [reply providers/jazzhr.mjs]: Went further than a patch here. Instead of chasing another pathname edge case, I re-derived board resolution to match the shape `providers/bamboohr.mjs` / `providers/breezy.mjs` / `providers/icims.mjs` already use for the same "subdomain = tenant, fixed canonical path" case: only the hostname is trusted, the input path is discarded entirely, and the board URL (`origin + /apply`) is always reconstructed rather than validated. `assertJazzHRUrl` dropped its pathname check for the same reason — host + HTTPS is the actual trust boundary (matches icims.mjs's own origin-only re-check on a posting href), so `/apply/D4/...` vs `/apply/` vs a stray unrelated path is no longer a distinction that function makes at all.

`resolvePostingUrl` — the content-recognition filter that decides which same-origin links look like a real posting during list parsing, not a trust boundary — still needed the fix you're pointing at: `.+` after `/apply/` matches a lone `/` too, so `/apply//` passed. Tightened to require the first character after `/apply/` not itself be a slash. Added a regression test and re-verified against all 5 live tenants — same job counts as every prior round.

File: providers/jazzhr.mjs [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+
+/** Resolve the tenant origin (`https://<tenant>.applytojob.com`) from an
+ * entry — honours an explicit `api:` URL, else `careers_url`. Mirrors
+ * providers/bamboohr.mjs / providers/breezy.mjs / providers/icims.mjs. */
+function resolveOrigin(entry) {
+  for (const raw of [entry?.api, entry?.careers_url]) {
+    if (typeof raw !== 'string' || !raw.trim()) continue;
+    let parsed;
+    try { parsed = new URL(raw.trim()); } catch { continue; }
+    if (parsed.protocol !== 'https:' || !HOST_RE.test(parsed.hostname)) continue;
```
[the remark this answers is not in the document]
Comment (ilyakanevskiy) [reply providers/jazzhr.mjs:57]: Confirmed and fixed — `resolveOrigin`/`assertJazzHRUrl` checked protocol and hostname but never port, and `resolvePostingUrl` compared `hostname` alone rather than the full origin. Not an SSRF vector on its own here (the hostname stays pinned to the real vendor domain either way), but inconsistent with treating origin as the actual trust boundary, which is the whole point of the prior refactor. Both now reject a non-default port, and `resolvePostingUrl` compares `url.origin` against `base.origin` (which also folds the protocol check in, since `base.origin` is always https). Added a regression test for `:8443` and re-verified against all 5 live tenants — same job counts.

[the remark this answers is not in the document]
Comment (ilyakanevskiy) [reply providers/jazzhr.mjs:57]: Follow-up: went one step further than validating the port — dropped the check entirely. `resolveOrigin` now rebuilds `https://${hostname}` directly instead of returning `parsed.origin`, matching `providers/bamboohr.mjs` / `providers/breezy.mjs` exactly. A port has nowhere to survive to once the origin is reconstructed from the hostname alone, so there's nothing left to check or reject — same logic already applied to the path. `assertJazzHRUrl`'s port check came out too, since `job.url` is always derived from `resolvePostingUrl`, which is already pinned to that portless origin.

**Comment by santifer** [comment]
Thanks for turning these around so fast, @ilyakanevskiy: I went back over the provider at your latest head and reproduced the `/apply//` case independently. It's correctly rejected now, and so is the board-root form with a query string.

I also chased down the case that worried me most: a healthy but genuinely empty board whose only apply link is its own root. That does throw rather than return an empty list. But following it through `scan.mjs`, the throw is caught per company and reported as a board error, so it never aborts a scan, and your two regression tests pin exactly that intent. So that's a deliberate call, not a bug: leaving it alone.

One thing worth fixing: the file header still says the permalink fallback runs "if that finds nothing". It runs unconditionally now, as your own inline comment and the partial-redesign test both spell out. One line, but it's the first thing the next reader will see.

After that the branch needs a refresh: main has moved on since your last push, so this will need CI on the new head before it can go further. One heads-up on sequencing: #3748 also edits the provider list comment in `templates/portals.example.yml`, so whichever of the two lands second will want a trivial rebase there.

**Comment by ilyakanevskiy** [comment]
Hey @santifer — fixed the header comment (it now matches what the inline comment and the partial-redesign test already said), and rebased onto the current `main` since it had moved on — new head is `dc552831`, CI should pick it up fresh. Thanks for the heads-up on #3748, I'll watch for it and rebase again if it lands first.
